### ISSUE TITLE
‘TimeOfDay’ not found compilation error

### ISSUE THREAD

**Issue by xiaoyu-nus** [dart-lang/sdk#42173]
I am using a time picker for my flutter web app running on chrome. I am using the BLOC logic. TimeOfDay is imported from time.dart in the bloc file and used by the event file:

The bloc file:
```
import 'dart:async';

import 'package:bloc/bloc.dart';
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';

part 'dispatch_event.dart';
part 'dispatch_state.dart';

class DispatchBloc extends Bloc<DispatchEvent, DispatchState> {
  @override
  DispatchState get initialState => DispatchInitial();

  @override
  Stream<DispatchState> mapEventToState(
    DispatchEvent event,
  ) async* {
    if (event is PickDispatchDateEvent) {
      yield PickDispatchDateState(event.date);
    } else if (event is PickDispatchTimeEvent) {
      yield PickDispatchTimeState(event.time);
    }
  }
}
```
Here is the event file:
```
part of 'dispatch_bloc.dart';

abstract class DispatchEvent extends Equatable {
  const DispatchEvent();
}

class PickDispatchDateEvent extends DispatchEvent {
  final DateTime date;
  PickDispatchDateEvent(this.date);
  @override
  List<Object> get props => [date];
}

class PickDispatchTimeEvent extends DispatchEvent {
  final TimeOfDay time;
  PickDispatchTimeEvent(this.time);
  @override
  List<Object> get props => [time];
}
```
However, when I run 
```
flutter run -d web-server --verbose
```
The following errors are thrown:
```
                    Compiler message:
[   +2 ms] lib/blocs/dispatch_bloc/dispatch_event.dart:15:9: Error: Type 'TimeOfDay' not found.
[   +3 ms]   final TimeOfDay time;
[   +2 ms]         ^^^^^^^^^
[   +5 ms] lib/blocs/dispatch_bloc/dispatch_state.dart:20:9: Error: Type 'TimeOfDay' not found.
[        ]   final TimeOfDay time;
[  +17 ms]         ^^^^^^^^^
[+7421 ms] lib/blocs/dispatch_bloc/dispatch_event.dart:15:9: Error: 'TimeOfDay' isn't a type.
[   +2 ms]   final TimeOfDay time;
[  +27 ms]         ^^^^^^^^^
[   +3 ms] lib/blocs/dispatch_bloc/dispatch_state.dart:20:9: Error: 'TimeOfDay' isn't a type.
[   +1 ms]   final TimeOfDay time;
[  +63 ms]         ^^^^^^^^^
[+50460 ms] Unhandled exception:
[   +6 ms] InvalidType doesn't have nullability
[   +2 ms] #0      InvalidType.nullability (package:kernel/ast.dart:6709:34)
[   +2 ms] #1      ProgramCompiler._mustBeNonNullable (package:dev_compiler/src/kernel/compiler.dart:3129:14)
[  +11 ms] #2      ProgramCompiler._emitArgumentInitializers.initParameter
(package:dev_compiler/src/kernel/compiler.dart:3163:18)
[ +171 ms] #3      ProgramCompiler._emitArgumentInitializers (package:dev_compiler/src/kernel/compiler.dart:3185:20)
[  +65 ms] #4      ProgramCompiler._emitConstructorBody (package:dev_compiler/src/kernel/compiler.dart:1375:16)
[   +5 ms] #5      ProgramCompiler._emitConstructor.<anonymous closure>.<anonymous closure>
           (package:dev_compiler/src/kernel/compiler.dart:1355:19)
[   +7 ms] #6      ProgramCompiler._superDisallowed (package:dev_compiler/src/kernel/compiler.dart:3122:24)
[   +6 ms] #7      ProgramCompiler._emitConstructor.<anonymous closure> (package:dev_compiler/src/kernel/compiler.dart:1354:15) 
[   +3 ms] #8      ProgramCompiler._withLetScope (package:dev_compiler/src/kernel/compiler.dart:2096:25)
[  +17 ms] #9      ProgramCompiler._withCurrentFunction (package:dev_compiler/src/kernel/compiler.dart:3112:18)
[  +17 ms] #10     ProgramCompiler._emitConstructor (package:dev_compiler/src/kernel/compiler.dart:1352:16)
[  +16 ms] #11     ProgramCompiler._defineConstructors (package:dev_compiler/src/kernel/compiler.dart:972:38)
[   +4 ms] #12     ProgramCompiler._emitClassDeclaration (package:dev_compiler/src/kernel/compiler.dart:567:19)
[   +2 ms] #13     ProgramCompiler._emitClass (package:dev_compiler/src/kernel/compiler.dart:510:21)
[   +3 ms] #14     List.forEach (dart:core-patch/growable_array.dart:282:8)
[   +2 ms] #15     ProgramCompiler._emitLibrary (package:dev_compiler/src/kernel/compiler.dart:459:23)
[   +8 ms] #16     List.forEach (dart:core-patch/growable_array.dart:282:8)
[   +4 ms] #17     ProgramCompiler.emitModule (package:dev_compiler/src/kernel/compiler.dart:345:15)
[  +21 ms] #18     JavaScriptBundler.compile (package:frontend_server/src/javascript_bundle.dart:133:33)
[  +41 ms] <asynchronous suspension>
[  +26 ms] #19     FrontendCompiler.writeJavascriptBundle (package:frontend_server/frontend_server.dart:627:20)
[  +15 ms] <asynchronous suspension>
[  +20 ms] #20     FrontendCompiler.compile (package:frontend_server/frontend_server.dart:533:15)
[  +15 ms] <asynchronous suspension>
[   +7 ms] #21     _FlutterFrontendCompiler.compile (package:flutter_frontend_server/server.dart:40:22)
[  +19 ms] #22     listenAndCompile.<anonymous closure> (package:frontend_server/frontend_server.dart:1164:26)
[  +10 ms] #23     _RootZone.runUnaryGuarded (dart:async/zone.dart:1374:10)
[   +6 ms] #24     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
[   +4 ms] #25     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:266:7)
[   +3 ms] #26     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:70:11)
[   +3 ms] #27     _EventSinkWrapper.add (dart:async/stream_transformers.dart:17:11)
[   +4 ms] #28     _StringAdapterSink.add (dart:convert/string_conversion.dart:238:11)
[  +12 ms] #29     _LineSplitterSink._addLines (dart:convert/line_splitter.dart:152:13)
[   +2 ms] #30     _LineSplitterSink.addSlice (dart:convert/line_splitter.dart:127:5)
[   +3 ms] #31     StringConversionSinkMixin.add (dart:convert/string_conversion.dart:165:5)
[   +3 ms] #32     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:122:24)
[  +10 ms] #33     _RootZone.runUnaryGuarded (dart:async/zone.dart:1374:10)
[   +2 ms] #34     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
[   +1 ms] #35     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:266:7)
[   +3 ms] #36     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:70:11)
[   +6 ms] #37     _EventSinkWrapper.add (dart:async/stream_transformers.dart:17:11)
[   +1 ms] #38     _StringAdapterSink.add (dart:convert/string_conversion.dart:238:11)
[   +3 ms] #39     _StringAdapterSink.addSlice (dart:convert/string_conversion.dart:243:7)
[   +3 ms] #40     _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:315:20)
[   +3 ms] #41     _Utf8ConversionSink.add (dart:convert/string_conversion.dart:308:5)
[   +5 ms] #42     _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:74:18)
[   +2 ms] #43     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:122:24)
[   +2 ms] #44     _RootZone.runUnaryGuarded (dart:async/zone.dart:1374:10)
[   +3 ms] #45     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
[  +10 ms] #46     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:266:7)
[   +2 ms] #47     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:779:19)
[   +4 ms] #48     _StreamController._add (dart:async/stream_controller.dart:655:7)
[   +7 ms] #49     _StreamController.add (dart:async/stream_controller.dart:597:5)
[   +2 ms] #50     _Socket._onData (dart:io-patch/socket_patch.dart:1999:41)
[   +3 ms] #51     _RootZone.runUnaryGuarded (dart:async/zone.dart:1374:10)
[   +2 ms] #52     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
[   +8 ms] #53     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:266:7)
[  +10 ms] #54     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:779:19)
[   +4 ms] #55     _StreamController._add (dart:async/stream_controller.dart:655:7)
[  +12 ms] #56     _StreamController.add (dart:async/stream_controller.dart:597:5)
[  +10 ms] #57     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1544:33)
[   +9 ms] #58     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1036:14)
[   +3 ms] #59     _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
[   +2 ms] #60     _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
[   +3 ms] #61     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
[   +7 ms] #62     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)
[  +69 ms] "flutter run" took 63,884ms.
the Dart compiler exited unexpectedly.

#0      throwToolExit (package:flutter_tools/src/base/common.dart:14:3)
#1      DefaultResidentCompiler._compile.<anonymous closure> (package:flutter_tools/src/compile.dart:710:13)
#2      _rootRun (dart:async/zone.dart:1180:38)
#3      _CustomZone.run (dart:async/zone.dart:1077:19)
#4      _CustomZone.runGuarded (dart:async/zone.dart:979:7)
#5      _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:392:13)
#6      _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:402:15)
#7      _BufferingStreamSubscription._close (dart:async/stream_impl.dart:286:7)
#8      _SinkTransformerStreamSubscription._close (dart:async/stream_transformers.dart:98:11)
#9      _EventSinkWrapper.close (dart:async/stream_transformers.dart:25:11)
#10     _StringAdapterSink.close (dart:convert/string_conversion.dart:251:11)
#11     _LineSplitterSink.close (dart:convert/line_splitter.dart:136:11)
#12     _SinkTransformerStreamSubscription._handleDone (dart:async/stream_transformers.dart:143:24)
#13     _rootRun (dart:async/zone.dart:1180:38)
#14     _CustomZone.run (dart:async/zone.dart:1077:19)
#15     _CustomZone.runGuarded (dart:async/zone.dart:979:7)
#16     _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:392:13)
#17     _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:402:15)
#18     _BufferingStreamSubscription._close (dart:async/stream_impl.dart:286:7)
#19     _SinkTransformerStreamSubscription._close (dart:async/stream_transformers.dart:98:11)
#20     _EventSinkWrapper.close (dart:async/stream_transformers.dart:25:11)
#21     _StringAdapterSink.close (dart:convert/string_conversion.dart:251:11)
#22     _Utf8ConversionSink.close (dart:convert/string_conversion.dart:303:20)
#23     _ConverterStreamEventSink.close (dart:convert/chunked_conversion.dart:83:18)
#24     _SinkTransformerStreamSubscription._handleDone (dart:async/stream_transformers.dart:143:24)
#25     _rootRun (dart:async/zone.dart:1180:38)
#26     _CustomZone.run (dart:async/zone.dart:1077:19)
#27     _CustomZone.runGuarded (dart:async/zone.dart:979:7)
#28     _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:392:13)
#29     _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:402:15)
#30     _BufferingStreamSubscription._close (dart:async/stream_impl.dart:286:7)
#31     _SyncStreamControllerDispatch._sendDone (dart:async/stream_controller.dart:787:19)
#32     _StreamController._closeUnchecked (dart:async/stream_controller.dart:644:7)
#33     _StreamController.close (dart:async/stream_controller.dart:637:5)
#34     _Socket._onData (dart:io-patch/socket_patch.dart:2006:21)
#35     _rootRunUnary (dart:async/zone.dart:1196:13)
#36     _CustomZone.runUnary (dart:async/zone.dart:1085:19)
#37     _CustomZone.runUnaryGuarded (dart:async/zone.dart:987:7)
#38     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
#39     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:266:7)
#40     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:779:19)
#41     _StreamController._add (dart:async/stream_controller.dart:655:7)
#42     _StreamController.add (dart:async/stream_controller.dart:597:5)
#43     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1551:35)
#44     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1029:18)
#45     _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
#46     _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
#47     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#48     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)



[  +97 ms] Syncing files to device Web Server... (completed in 62,231ms, longer than expected)
[  +77 ms] Synced 0.0MB.
[  +40 ms] Failed to compile application.
```
In the component class where the time picker is used, the type TimeOfDay is also used but I dont see any error. Here is the code segment:
```
Future<TimeOfDay> getTime(context) {
    // Imagine that this function is
    // more complex and slow.
    return showTimePicker(
      context: context,
      initialTime: TimeOfDay.fromDateTime(DateTime(2020, 1, 1, 8, 30)),
      builder: (BuildContext context, Widget child) {
        return Theme(
          data: ThemeData.light(),
          child: child,
        );
      },
    );
  }

  void callTimePicker(context) async {
    var time = await getTime(context);
    _dispatchBloc.add(PickDispatchTimeEvent(time));
  }
```
Here is my doctor output:
```
[√] Flutter (Channel beta, 1.18.0-11.1.pre, on Microsoft Windows [Version 10.0.18362.836], locale en-SG)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[√] Chrome - develop for the web
[√] Android Studio (version 3.6)
[!] IntelliJ IDEA Ultimate Edition (version 2019.2)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[√] VS Code (version 1.45.1)
[√] Connected device (2 available)

! Doctor found issues in 1 category.
```

My dependencies:
```
version: 1.0.0+1

environment:
  sdk: ">=2.7.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  charts_flutter:
  flutter_bloc:
  flutter_staggered_grid_view:
  form_field_validator:
  validators:
  equatable:
  universal_html:
  fl_chart:
  flutter_datetime_picker:
  intl:

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.3

dev_dependencies:
  flutter_test:
    sdk: flutter
```

My VS code doesn't show the red squiggle lines for TimeOfDay and when I ctrl + click it directs me to the time.dart file. However the compiler doesn't seems to recognise it. Please help me with the issue.

**Comment by lrhn** [comment]
> TimeOfDay is imported from time.dart in the bloc file

I don't see any import of `time.dart` in the bloc file you displayed above. Are you sure it's there?

**Comment by xiaoyu-nus** [comment]
I think import `package:flutter/material.dart` includes `time.dart` doesn't it.

**Comment by lrhn** [comment]
It does appear to do so: https://api.flutter.dev/flutter/material/TimeOfDay-class.html
(I'm clearly not a Flutter expert).

As such, the type should be available in the part files included in the library importing `material.dart`. So, it is perhaps a bug in the front-end.

**Comment by johnniwinther** [comment]
@jensjoha Can you repro this?

**Comment by jensjoha** [comment]
I unfortunately cannot.

**Comment by johnniwinther** [comment]
Given that failing to find a type that _is_ defined in another library is at the core of the compilation it is unlikely that the error is in the front-end itself (virtual all compilations would fail if that was the case).

This leads me to believe that it is something in the local environment/cache that has gone wrong. @xiaoyu-nus can you reproduce the failure after doing a `flutter clean` or if you create the source code from scratch in a new project?

**Comment by xiaoyu-nus** [comment]
@johnniwinther Thanks for the help! `flutter clean` which doesn't work. I then did a `git clone` from my repo to another directory and suddenly all of the packages import statement in the entire project fail, including local packages. `flutter packages get` or `flutter pub get` doesn't work either.

![image](https://user-images.githubusercontent.com/51201915/84012061-91dd7b80-a9a9-11ea-9936-6e4379e83fd7.png)

Other non package import works fine. I am not sure what's wrong or missing.

**Comment by johnniwinther** [comment]
This sounds like the packages are not configured correctly or at all. Do
you have a .dart_tool/package_config.json file in you setup? I think `pub
get` or `pub update` should normally create this.

**Comment by johnniwinther** [comment]
@xiaoyu-nus Any update on this?

**Comment by xiaoyu-nus** [comment]
Hi, thank you for the suggestions. The issue occured for a second time on a local class which seems completely random. But after I restarted the editor several times (after running pub get etc) the problem is gone. The commands probably takes effect only after a restart so I dont know which one solved the issue. I waited to see whether it would come out again but it didn't. I guess you are right that it's related to local environment.
